galley/pkg/source/kube/source.go

```diff
@@ -0,0 +1,149 @@
+	"istio.io/istio/galley/pkg/source/kube/client"
+	"istio.io/istio/galley/pkg/source/kube/dynamic"
+	kubeConverter "istio.io/istio/galley/pkg/source/kube/dynamic/converter"
+	"istio.io/istio/galley/pkg/source/kube/log"
+	"istio.io/istio/galley/pkg/source/kube/schema"
+
+	kubeDynamic "k8s.io/client-go/dynamic"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	// TODO(nmittler): Pilot allows specifying the namespace for the shared informer.
+	namespace = ""
+)
+
+// New creates a new kube source for the given schema.
+func New(interfaces client.Interfaces, resyncPeriod time.Duration, schema *schema.Instance,
+	cfg *kubeConverter.Config) (runtime.Source, error) {
+
+	var err error
+	var cl kubernetes.Interface
+	var dynClient kubeDynamic.Interface
+	var sharedInformers informers.SharedInformerFactory
+
+	sources := make([]runtime.Source, 0)
+	for i, spec := range schema.All() {
+		log.Scope.Infof("[%d]", i)
+		log.Scope.Infof("  Source:      %s", spec.CanonicalResourceName())
+		log.Scope.Infof("  Collection:  %s", spec.Target.Collection)
+
+		// If it's a known type, use a custom (optimized) source.
+		if builtin.IsBuiltIn(spec.Kind) {
+			// Lazy create the kube client.
+			if cl, err = getKubeClient(cl, interfaces); err != nil {
+				return nil, err
+			}
+			sharedInformers = getSharedInformers(sharedInformers, cl, resyncPeriod, namespace)
+
+			source, err := builtin.NewSource(sharedInformers, spec)
+			if err != nil {
+				return nil, err
+			}
+			sources = append(sources, source)
+		} else {
+			// Lazy-create the dynamic client
+			if dynClient, err = getDynamicClient(dynClient, interfaces); err != nil {
+				return nil, err
+			}
+			// Unknown types use the dynamic source.
+			source, err := dynamic.NewSource(dynClient, resyncPeriod, spec, cfg)
+			if err != nil {
+				return nil, err
+			}
+			sources = append(sources, source)
+		}
+	}
+
+	return &aggregate{
+		sources: sources,
+	}, nil
+}
+
+func getKubeClient(cl kubernetes.Interface, interfaces client.Interfaces) (kubernetes.Interface, error) {
+	if cl != nil {
+		return cl, nil
+	}
+	return interfaces.KubeClient()
+}
+
+func getSharedInformers(sharedInformers informers.SharedInformerFactory,
+	cl kubernetes.Interface, resyncPeriod time.Duration, namespace string) informers.SharedInformerFactory {
+	if sharedInformers != nil {
+		return sharedInformers
+	}
+	return informers.NewSharedInformerFactoryWithOptions(cl, resyncPeriod, informers.WithNamespace(namespace))
+}
+
+func getDynamicClient(cl kubeDynamic.Interface, interfaces client.Interfaces) (kubeDynamic.Interface, error) {
+	if cl != nil {
+		return cl, nil
+	}
+	return interfaces.DynamicInterface()
+}
+
+type aggregate struct {
+	mu      sync.Mutex
+	sources []runtime.Source
+}
+
+func (s *aggregate) Start(handler resource.EventHandler) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Create an intermediate handler that captures sync events from each source.
+	syncGroup := sync.WaitGroup{}
```

Comment:
Minor Suggestion: You can use a plain integer and do atomic.* to count-down and then trigger the event in-line.  It removes the goroutine below.